spi/imx6ull-qspi/qspi.c

```diff
@@ -0,0 +1,326 @@
+
+#define QSPI_FLSHCR_ADDR (qspi_common.base + 0x0C / sizeof(uint32_t))
+
+#define QSPI_MCR_ADDR     (qspi_common.base)
+#define QSPI_MCR_CLR_TXF  (0x01 << 11)
+#define QSPI_MCR_CLR_RXF  (0x01 << 10)
+#define QSPI_MCR_CLR_MDIS (0x01 << 14)
+
+#define QSPI_BFGENCR_ADDR (qspi_common.base + 0x20 / sizeof(uint32_t))
+
+#define QSPI_SR_ADDR   (qspi_common.base + 0x15C / sizeof(uint32_t))
+#define QSPI_SR_BUSY   (0x01)
+#define QSPI_SR_RXWE   (0x01 << 16)
+#define QSPI_SR_TXEDA  (0x01 << 24)
+#define QSPI_SR_TXFULL (0x01 << 27)
+
+#define QSPI_RBDRn_ADRR(n) (qspi_common.base + (0x200 + 4 * (n)) / sizeof(uint32_t))
+
+#define QSPI_RBSR_ADRR    (qspi_common.base + 0x10C / sizeof(uint32_t))
+#define QSPI_RBSR_TO_READ (((*QSPI_RBSR_ADRR) >> 8) & 0x3F)
+
+#define QSPI_SFAR_ADDR (qspi_common.base + 0x100 / sizeof(uint32_t))
+
+#define QSPI_RBCT_ADDR  (qspi_common.base + 0x110 / sizeof(uint32_t))
+#define QSPI_RBCT_RXBRD (0x01 << 8)
+
+#define QSPI_FR_ADDR (qspi_common.base + 0x160 / sizeof(uint32_t))
+#define QSPI_FR_RBDF (0x01 << 16)
+
+#define QSPI_TBDR_ADDR (qspi_common.base + 0x154 / sizeof(uint32_t))
+
+#define QSPI_SFA1AD (qspi_common.base + 0x180 / sizeof(uint32_t))
+#define QSPI_SFA2AD (qspi_common.base + 0x184 / sizeof(uint32_t))
+#define QSPI_SFB1AD (qspi_common.base + 0x188 / sizeof(uint32_t))
+#define QSPI_SFB2AD (qspi_common.base + 0x18C / sizeof(uint32_t))
+
+#define WATERMARK 16
+
+#define NUM_LUT_SEQ 16
+
+typedef struct {
+	int mux;
+	char val;
+	char sion;
+} qspi_pctl_mux_t;
+
+typedef struct {
+	int pad;
+	char speed;
+	char dse;
+	char sre;
+} qspi_pctl_pad_t;
+
+qspi_pctl_mux_t qspi_pctl_mux[2][8] = {
+	{ { pctl_mux_nand_d7, 2, 0 }, { pctl_mux_nand_ale, 2, 0 }, { pctl_mux_nand_wp, 2, 0 }, { pctl_mux_nand_rdy, 2, 0 }, { pctl_mux_nand_ce0, 2, 0 }, { pctl_mux_nand_ce1, 2, 0 }, { pctl_mux_nand_cle, 2, 0 }, { pctl_mux_nand_dqs, 2, 0 } },
+	{ { pctl_mux_nand_re, 2, 0 }, { pctl_mux_nand_we, 2, 0 }, { pctl_mux_nand_d0, 2, 0 }, { pctl_mux_nand_d1, 2, 0 }, { pctl_mux_nand_d2, 2, 0 }, { pctl_mux_nand_d3, 2, 0 }, { pctl_mux_nand_d4, 2, 0 }, { pctl_mux_nand_d5, 2, 0 } },
+};
+
+qspi_pctl_pad_t qspi_pctl_pad[2][8] = {
+	{ { pctl_pad_nand_d7, 1, 4, 0 }, { pctl_pad_nand_ale, 1, 4, 0 }, { pctl_pad_nand_wp, 1, 4, 0 }, { pctl_pad_nand_rdy, 1, 4, 0 }, { pctl_pad_nand_ce0, 1, 4, 0 }, { pctl_pad_nand_ce1, 1, 4, 0 }, { pctl_pad_nand_cle, 1, 4, 0 }, { pctl_pad_nand_dqs, 1, 4, 0 } },
+	{ { pctl_pad_nand_re, 1, 4, 0 }, { pctl_pad_nand_we, 1, 4, 0 }, { pctl_pad_nand_d0, 1, 4, 0 }, { pctl_pad_nand_d1, 1, 4, 0 }, { pctl_pad_nand_d2, 1, 4, 0 }, { pctl_pad_nand_d3, 1, 4, 0 }, { pctl_pad_nand_d4, 1, 4, 0 }, { pctl_pad_nand_d5, 1, 4, 0 } },
+};
+
+
+struct {
+	volatile uint32_t *base;
+	uint32_t flash_base_addr[2];
+	bool init;
+} qspi_common = { .base = NULL, .init = false, .flash_base_addr = { QSPI_MMAP_BASE, QSPI_MMAP_BASE + QSPI_FLASH_A1_SIZE + QSPI_FLASH_A2_SIZE } };
+
+// TODO ustawić ile ohmów w rejestrze na flashu.
```

Comment:
Do not use any language other than english in the code, use C style comments
```suggestion
/* TODO set ohm value in the flash register */
```